### DISCUSSION TITLE
Callback function to trigger side-effects when server is ready

### DISCUSSION THREAD

**Discussion by habovh** [vercel/next.js#22145] [Ideas]
**Describe the feature you'd like to request**

I'm in the process of getting rid of my custom server that was mostly used with `next-routes`.

I managed to migrate my app to using Next's dynamic routing based on the filesystem, but I'm now stuck with the second thing the custom server did.

When the server is ready (listening for incoming requests), I need to trigger some side effects so my server knows Next is now able to serve requests.

I used to call a simple `fs.openSync('/tmp/app-initialized', 'w')` in my custom server to do just that, but I'm not sure how I can achieve this without one.

**Describe the solution you'd like**

I'm suggesting some kind of callback configurable in the `next.config.js`. Maybe a path to a script file or a function to run, I'm not sure what would be best.

**Describe alternatives you've considered**

I had a look into [Next server's code](https://github.com/vercel/next.js/blob/canary/packages/next/server/next-dev-server.ts) to see if there was already something like this alongside the `ready - started server on ...` console output with no luck.

As a workaround I'm currently calling a custom script that runs after `next build` and right before `next start`, but it is not ideal as even if the server listens quickly, there still might be some dropped requests in the meantime.

For those wondering, I'm deploying to Heroku with the Nginx buildpack which in our case requires to touch `/tmp/app-initialized` to signal the server is now able to serve requests.

Any thoughts appreciated!

**Comment by habovh** [comment]
Really no one has a similar need? I think the feature would be rather easy to implement but I wouldn't like to live on a personal fork of the library just to be able to do this...

Anyone?

**Reply by stigkj** [reply]
That would be really nice to have for all deployment scenarios where the app won't get any traffic before it is ready, for example with Kubernetes, Docker Swarm, Mesos, Nomad, etc.

What I do now is to have an API endpoint that just returns an emty http response with status code 200. Then most of the Next.js machinery is ready, I think 😜

**Comment by habovh** [comment]
I'm pinging this discussion back up.

The solution presented by @stigkj is not ideal as it requires an additional software constantly pinging a special test endpoint. I'm even keen on doing this myself but I'd at least like some input from the community as to *how* to implement this.

We could define the path to an executable file in next config, or maybe the name of a npm/package.json script to run within the project, or maybe something totally different.

What do you think would be the best approach here to allow for an *"on server listening"* callback/custom logic?

**Reply by stigkj** [reply]
So Heroku does not support live/ready endpoints like Kubernetes, Docker Swarm, Nomad, etc have? I.e. you can specify an http endpoint that the deployment platform checks and only if it returns 200 it will make the instance available to users.

**Reply by habovh** [reply]
It's more an issue about one of our dependencies that with Heroku itself.
To be specific, Heroku considers your app ready to handle requests as soon as it binds to the PORT env variable that it was provided. In our case, we need to have an Nginx process acting as a proxy in front of our app. So instead of our app binding to the PORT env variable provided by Heroku, it binds to a custom port that we configure our Nginx to forward requests to. In order to have Nginx running in front of our app, we use the [nginx buildpack](https://github.com/heroku/heroku-buildpack-nginx), which takes care of including nginx in our slug. However since Nginx does not know anything about our app, it needs a way to know when to launch and bind the PORT environment variable provided by Heroku. Launching the nginx process right away would trick Heroku into believing the app is ready when our app is actually still booting. In this case, Heroku might send requests to the booting slug and these would get a timeout between our nginx proxy and the actual app.

**TL;DR**

It's a specific use case we're having here, but in my opinion it still makes sense to have some kind of way to "hook" scripts onto app lifecycle events, such as "app is listening".
